protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,545 @@
+        // Expect final result to contain peer from each disjoint path, even though not all are
+        // among the best ones.
+        assert!(final_peers.contains(malicious_response_1[0].preimage()));
+        assert!(final_peers.contains(response_2[0].preimage()));
+        assert!(final_peers.contains(response_3[0].preimage()));
+    }
+
+    fn random_peers(n: usize) -> impl Iterator<Item = PeerId> + Clone {
+        (0 .. n).map(|_| PeerId::random())
+    }
+
+    #[derive(Debug, Clone)]
+    struct Graph(HashMap<PeerId, Peer>);
+
+    impl Arbitrary for Graph {
+        fn arbitrary<G: Gen>(g: &mut G) -> Self {
+            let mut peers = HashMap::new();
+            let mut peer_ids = random_peers(g.gen_range(K_VALUE.get(), 500))
+                .collect::<Vec<_>>();
+
+            for peer_id in peer_ids.clone() {
+                peer_ids.shuffle(g);
+
+                peers.insert(peer_id, Peer{
+                    known_peers: peer_ids[0..g.gen_range(K_VALUE.get(), peer_ids.len())].to_vec(),
+                });
+            }
+
+            Graph(peers)
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    struct Peer {
+        known_peers: Vec<PeerId>,
+    }
+
+    #[derive(Debug, Clone)]
+    struct Parallelism(usize);
+
+    impl Arbitrary for Parallelism{
+        fn arbitrary<G: Gen>(g: &mut G) -> Self {
+            Parallelism(g.gen_range(1, 10))
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    struct NumResults(usize);
+
+    impl Arbitrary for NumResults{
+        fn arbitrary<G: Gen>(g: &mut G) -> Self {
+            NumResults(g.gen_range(1, K_VALUE.get()))
+        }
+    }
+
+    enum PeerIterator {
+        Disjoint(ClosestDisjointPeersIter),
+        Closest(ClosestPeersIter),
+    }
+
+    impl PeerIterator {
+        fn next(&mut self, now: Instant) -> PeersIterState {
+            match self {
+                PeerIterator::Disjoint(iter) => iter.next(now),
+                PeerIterator::Closest(iter) => iter.next(now),
+            }
+        }
+
+        fn on_success(&mut self, peer: &PeerId, closer_peers: Vec<PeerId>) {
+            match self {
+                PeerIterator::Disjoint(iter) => iter.on_success(peer, closer_peers),
+                PeerIterator::Closest(iter) => iter.on_success(peer, closer_peers),
+            }
+        }
+
+        fn into_result(self) -> Vec<PeerId> {
+            match self {
+                PeerIterator::Disjoint(iter) => iter.into_result().collect(),
+                PeerIterator::Closest(iter) => iter.into_result().collect(),
+            }
+        }
+    }
+
+    /// Ensure [`ClosestPeersIter`] and [`ClosestDisjointPeersIter`] yield similar closest peers.
+    ///
+    // NOTE: One can not ensure both iterators yield the *same* result. While [`ClosestPeersIter`]
+    // always yields the closest peers, [`ClosestDisjointPeersIter`] might not. Imagine a node on
+    // path x yielding the 22 absolute closest peers, not returned by any other node. In addition
```

Comment:
> [..] yielding the 22 absolute closest peers, _not returned by any other node_.

That is a strange assumption though. In a non-adversarial setting that seems unlikely since all disjoint paths converge to the same closest peers by the nature of the iterative query process, only differing in which subsets of these are queried on each path. Thus any "unused" peers on one path should still be seen (and can be queried) on other paths if they are indeed closer than others. So, I still don't see how, in a non-adversarial setting with perfect network connectivity (e.g. in this test setup), the results could possibly diverge? Could you elaborate?

Reply:
> In this test I don't want to assume a well connected graph and rather as well test the edge cases.

But the weaker the assertions one can make in a test, the less useful the test, because it will accept more implementations (including incorrect ones). Now I'm not saying that it may not be useful to check some properties of the algorithm in less ideal conditions, as long as one can still make some useful assertions, but the reason I suggested the test for exact agreement of the result with the standard Kademlia lookup in the absence of connectivity problems and adversarial behaviour is that it establishes a baseline for correctness: If an implementation does not pass that, there is little hope it will work correctly under less ideal conditions. And the reason I expect agreement of these algorithms on the result is that, as far as my understanding goes, every path in a multi-path lookup should essentially end up seeing the same closest peers to the key being looked up, given that each peer has full knowledge of its close neighbourhood in the overlay (routing table) - again in ideal conditions of network connectivity (but not necessarily assuming every node knows every other, only that every node can reach every other via some path in the overlay routing) and no adversarial nodes in the query.

Reply:
> So, I still don't see how, in a non-adversarial setting with perfect network connectivity (e.g. in this test setup), the results could possibly diverge? Could you elaborate?

After trying to adjust the test accordingly I have found the following edge case where a `ClosestPeersIter` and a `ClosestDisjointPeersIter` yield different results running on a fully connected graph.

### Example

There are 21 (1, 2, ..., 21) remote peers total (excluding the local node). Each node is aware of all other nodes (fully connected). Both `ClosestPeersIter` as well as `ClosestDisjointPeersIter` are initialized with all but the closest nodes (2, 3, ..., 21). The amount of allowed parallelism and thus the amount of disjoint paths is `3`. `num_results` is `3` as well (one per disjoint path).

(This is the simplest example I can come up with. More complex ones are possible.)

#### `ClosestPeersIter`

- Query node `2` (the closest node the iterator is aware of) and thus learn of node `1`.

- Query node `1`, but don't learn anything new.

- Query node `3`, but don't learn anything new.

- **Return nodes `1`, `2`, `3`.**

#### `ClosestDisjointPeersIter`

- Path `1`: Query node `2` (the closest node the iterator is aware of) and thus learn of node `1`.

- Path `2`: Try query node `2`, which was already queried by path `1` and thus `on_failure` is called.

- Path `2`: Query node `3` and thus learn of node `1`.

- Path `3`: Try query node `2`, which was already queried by path `1` and thus `on_failure` is called.

- Path `3`: Try query node `3`, which was already queried by path `2` and thus `on_failure` is called.

- Path `3`: Query node `4` and thus learn of node `1`.

- Path `1` Query node `1`, but don't learn anything new. (Path `1` is finished yielding node `1`.)

- Path `2`: Try query node `1`, which was already queried by path `1` and thus `on_failure` is called. (Path `2` is finished yielding node `3`.)

- Path `3`: Try query node `1`, which was already queried by path `1` and thus `on_failure` is called. (Path `3` is finished yielding node `4`.)

- **Return nodes `1`, `3`, `4`.**

In other words: The first path of the `ClosestDisjointPeersIter` tries node `2` and thus finds out about node `1` which it eventually yields. Given that it queried node `2` no other path can query / yield node `2` and thus the final output misses node `2` which the output of a `ClosestPeersIter` would not.

Does the above make sense @romanb?

While I will put more thoughts into this, I would suggest moving forward requiring both iterators to find the single closest nodes and requiring that the two results overlap like @infinity0 suggested above.

> @mxinden It is sufficient to test that with the multipaths approach, the result of the lookup, which is a set of nodes, contains the actual closest node - which you know, because you are defining the test.

> A slightly more strict test would check that the result of the lookup, is approximately equal (say > >1/2 overlap) with the sibling nodes that S/Kademlia additionally keeps. I am not sure if you > implemented this in your PR however.

Reply:
Thanks for the example. It does show that I overlooked an important detail of the current implementation. Let me try to summarise in my own words: We may not end up with all the actual `k` closest nodes to the key, because any one path may contact peers that are among the `k` closest but not among the `k / d` closest that this path is allowed to contribute to the overall `k` results. Thinking then about what guarantees we actually have, it seems that we can only expect at least `1 / d` of the resulting list of peers to be among the actual `k` closest (according to the non-disjoint query). That kind of dispersion in a completely non-adversarial setting does not really seem acceptable and is unlikely to be intended by the S/Kademlia paper, as it directly affects where records are stored and whether they are found.

Intuitively, the problem with this approach then is that every individual path is allowed to "race arbitrarily far ahead" towards the `k` closest peers, thus contacting these but then only contributing its `k / d` fraction of them (naturally as permitted, since no path should be allowed to make a larger contribution to the resulting list of peers than any other).

A first improvement that suggests itself is to enforce fairness in the progression across all paths, i.e. a sort of lock-step execution where, after a path has contacted a new peer, it may only contact another after all other paths have contacted one more peer (or terminated). That would guarantee, I think, an overlap of `k - (k / d)` peers among all lookups for that key in a non-adversarial setting. Still, that is not great. I really think that in the non-adversarial setting the disjoint-path lookup should guarantee to find all of the "true"  `k` closest peers to the key.

So I went back to the paper and read "[..] By using the sibling list [..] the lookup doesn’t converge at a single node but terminates on d close-by neighbors, which all know the complete siblings for the destination key." I read this such that as long as at least one path is free of adversarial nodes, the complete list of "true" siblings of the key should be discovered by the query. To realise that in the absence of special handling for sibling lists, I've had the following thought: When a path `a` wants to contact peer `p`, but `p` is already being contacted by another path `b`, then instead of always regarding `p` as "failed" for `a`, it is instead reported to `a` with the same outcome as for `b` (success or failure), however _without reporting to `a` the supposed closer peers obtained from `p` in case of success (these are only reported to `b`)_. Thus `a` does not incorporate any peers from other paths into its own, but it still recognises a successful contact made on another path to a peer _that this path would have liked to query as well (because it obtained that peer earlier on in this same path)_.  Furthermore, each path must then collect a full `k` results. When all paths have terminated, the `k` results from all paths are merged into one set. In the completely non-adversarial setting (and no connectivity problems), all these results will be the same and the resulting set of size `k` contains the "true" `k` closest peers. In an adversarial setting and / or with partial network connectivity between some of the nodes, the resulting set will be larger than `k` and if this set is supposed to be used in a follow-up `STORE` operation (e.g. in the context of `put_record` or `add_provider`), that request is nevertheless sent to this full set of peers, which may be more than `k`, but we are then assured that an attempt is made to send it to the "true" `k` closest peers to the key. 

Thoughts?

Reply:
> When a path a wants to contact peer p, but p is already being contacted by another path b, then instead of always regarding p as "failed" for a, it is instead reported to a with the same outcome as for b (success or failure), however without reporting to a the supposed closer peers obtained from p in case of success (these are only reported to b). 

What's the point of marking it as "success" if you're not going to use it?

>  I read this such that as long as at least one path is free of adversarial nodes, the complete list of "true" siblings of the key should be discovered by the query.

Bear in mind that it is impossible to actually know which one is the adversary, the best thing we can do is to keep the width of the multipath as wide as possible.

Reply:
> What's the point of marking it as "success" if you're not going to use it?

Marking it as "success" allows the path to return the peer in its final result.

Reply:
> Marking it as "success" allows the path to return the peer in its final result.

And what is the point of that? If the query to this peer succeeded this implies we already got some nodes X closer than this peer; it is then a bit pointless to keep this peer in the final result, who is just going to query X again.

Reply:
Let me adjust the [example above](https://github.com/libp2p/rust-libp2p/pull/1473#discussion_r421771743) according to Roman's suggestion to better explain the idea.

There are 21 (1, 2, ..., 21) remote peers total (excluding the local node). Each node is aware of all other nodes (fully connected). Both `ClosestPeersIter` as well as `ClosestDisjointPeersIter` are initialized with all but the closest nodes (2, 3, ..., 21). The amount of allowed parallelism and thus the amount of disjoint paths is `3`. `num_results` is `3` as well (one per disjoint path).

(This is the simplest example I can come up with. More complex ones are possible.)

#### `ClosestPeersIter`

- Query node `2` (the closest node the iterator is aware of) and thus learn of node `1`.

- Query node `1`, but don't learn anything new.

- Query node `3`, but don't learn anything new.

- **Return nodes `1`, `2`, `3`.**

#### `ClosestDisjointPeersIter`

- Path `1`: Query node `2` (the closest node the iterator is aware of) and thus learn of node `1`.

- Path `2`: Try query node `2`, which was already queried by path `1` and thus `on_failure` is called.

- Path `2`: Query node `3` and thus learn of node `1`.

- Path `3`: Try query node `2`, which was already queried by path `1` and thus `on_failure` is called.

- Path `3`: Try query node `3`, which was already queried by path `2` and thus `on_failure` is called.

- Path `3`: Query node `4` and thus learn of node `1`.

- Path `1` Query node `1`, but don't learn anything new. (Path `1` is finished yielding node `1`.)

- Path `2`: Try query node `1`, which was already queried by path `1` and thus `on_failure` is called. (Path `2` is finished yielding node `3`.)

- Path `3`: Try query node `1`, which was already queried by path `1` and thus `on_failure` is called. (Path `3` is finished yielding node `4`.)

- **Return nodes `1`, `3`, `4`.**

#### `ClosestDisjointPeersIter` with Roman's suggestion

- Path `1`: Query node `2` (the closest node the iterator is aware of) and thus learn of node `1`.

- Path `2`: Try query node `2`, which was already queried by path `1` and thus `on_success` is called without any closer peers. Path 2 is allowed to go again.

- Path `2`: Query node `3` and thus learn of node `1`.

- Path `3`: Try query node `2`, which was already queried by path `1` and thus `on_success` is called without any closer peers. Path 3 is allowed to go again.

- Path `3`: Try query node `3`, which was already queried by path `2` and thus `on_success` is called without any closer peers. Path 3 is allowed to go again.

- Path `3`: Query node `4` and thus learn of node `1`.

- Path `1` Query node `1`, but don't learn anything new. (Path `1` is finished yielding node `1`.)

- Path `2`: Try query node `1`, which was already queried by path `1` and thus `on_success` is called without any closer peers. Path 2 is allowed to go again, but given that it is not aware of any closer nodes, it finishes.

- Path `3`: Try query node `1`, which was already queried by path `1` and thus `on_success` is called without any closer peers. Path 3 is allowed to go again, but given that it is not aware of any closer nodes it finishes.

- Collect the closest nodes from each path ({1, 2, 3}, {1, 2, 3}, {1, 2, 3}) and collect `num_results` unique peers (in this case 3) in a round robin fashion from the paths. **Return `1`, `2`, `3`.

Reply:
> What's the point of marking it as "success" if you're not going to use it?

I simply overlooked that the naive splitting of the total number of desired results among the paths of a query as was done here so far is not a good idea, because it easily leads to a large dispersion in the "closest peers" reported for a key even in a perfect setting (no adversarial nodes, perfect connectivity), as described above. My last suggestion is essentially just a patch to correct that by allowing each path to return the full number of requested results and then merging them. That means in an ideal setting all paths must yield the same results. By marking a successfully contacted peer as a "success" on all paths that encountered this peer if one of them successfully queries it (but still only allowing one path to make use of the reported closer peers to preserve disjointness), then in the absence of adversarial nodes or partial connectivity, each path will indeed yield the same results - the result set size only grows beyond the requested number if at least one path contains adversarial nodes and / or at least one path has connectivity issues to some peers that prevent it from getting to know all the closest peers to the desired key. Of course natural churn of nodes joining and leaving while the query is in progress can have similar effects on the dispersion of the results. In any case, for a subsequent store operation based on this peer result set, by always performing it on this full set of peers, as long as there is one successful path (i.e. path without any adversarial nodes) the store operation gets sent to all the "correct" peers, since, as you rightly say, we naturally don't know which of the paths was successful, if any, and thus which of the resulting peers are genuine.

Reply:
> 

> ClosestPeersIter
> Return nodes 1, 2, 3.
> ClosestDisjointPeersIter
> Return nodes 1, 3, 4.
> ClosestDisjointPeersIter with Roman's suggestion
> Return nodes 1, 2, 3.

In S-Kademlia the d nodes closest to

Reply:
> are initialized with all but the closest nodes (2, 3, ..., 21)

I don't understand what this means, Kademlia and S-Kademlia only talk about querying for a single key not a set of keys at once. Also what ID is the local node, 0?

> ClosestPeersIter
> Return nodes 1, 2, 3.
> ClosestDisjointPeersIter
> Return nodes 1, 3, 4.
> ClosestDisjointPeersIter with Roman's suggestion
> Return nodes 1, 2, 3.

In S-Kademlia the d nodes closest to a given key are supposed to store values for the key, so I'm having a hard time understanding your example. So for example if d = 3 then when querying for `1` your result set should be `{0, 1, 3}` since their XOR distance to the query `1` is `(0, 1, 2)` respectively.

But even ignoring that, your example seems consistent with what I said above - since 2, 3, and 4 all previously gave you 1 as a "next hop candidate" for whatever reason, it would be redundant to return `(1, 2, 3)` or `(1, 3, 4)` as the result of the final query, either set is fine, the result 1 is what's important. Also `num_results` for S-Kademlia should just be automatically the same as `d`, to simplify things. If you get < d results from the final hop, I guess you *could* return "closer" nodes from previous hops to "fill up" d results, but IMO it is a bit pointless since they already told you they don't have the key you're querying for, and gave you other nodes to contact.

Generally speaking, I think things make much more sense when trying to understand the algorithm from the multipath perspective, not the disjoint paths perspective.

Reply:
> I guess you could return "closer" nodes from previous hops to "fill up" d results, but IMO it is a bit pointless [..]

Actually, I take that back, since you'll need this for inserting values. Let me think about the example for a bit.

Reply:
OK, I think I understand the problem now. It is similar to the problem I described [here](https://forum.web3.foundation/t/s-kademlia-and-multipaths/310) that motivated the multipaths perspective, with an additional problem that you really do want to return the closest set of nodes.

I missed this problem in my original multipaths based solution, but it is easily fixable - if you have < d results, you keep taking results from the previous hop, and so on:

~~~~
hop1 hop2 hop3

3 -> 2 -> 1
     ^
4 ---|
     ^
5 ---|
~~~~

If you want 3 results, you walk the graph backwards on a per-hop basis, starting from hop3, getting {1, 2, 3}.

However with the disjoint paths strategy, you have result-sets-per-path instead. From my example above:

- path 1: {3, 2, 1}
- path 2: {4, 2, 1}
- path 3: {5, 2, 1}

If you simply "round robin" across these result sets, starting from the first one, you will end up with something like {1, 2, 5} which is incorrect. In order to get the correct result via round-robin, you have to somehow know in advance that the first result set must be looked at last. The simplest way to do this correctly is to re-arrange these result sets into a graph, and traverse the graph like I showed above. Or, something equivalent to that approach.

(note that naive attempts to fix the round-robin algorithm without constructing a graph, won't be able to handle more complex situations, like:

- {4, 2, 1}
- {5, 2, 1}
- {6, 3, 2, 1}
- {7, 2, 1}
- {8, 2, 1}

In order to write a general algorithm to unbury that 3 *whilst retaining the disjointness property*, you need to track not "which ones the round robin has visited" (since the order is arbitrary and can screw up the logic) but rather all of the edges in the graph of queries.

So hopefully this is another demonstration why the multipaths viewpoint is much more natural.

Reply:
Actually, this more complex case breaks all of the suggested solutions so far, including my "easy fix" to the multipaths approach:

~~~~
hop1 hop2 hop3

3 -> 2 -> 1
     ^    ^
4 ---|    |
     |    |
5 ---|    |
          |
7 -> 6 ---|
          |
9 -> 8 ---|
~~~~

My suggested way of proceeding backwards on a per-hop basis will give {1, 2, 6, 8, 3} but we want {1, 2, 3, 4, 5}...

Reply:
On second thoughts, the "easy fix" solution of {1, 2, 6, 8, 3} is indeed correct because {1, 2, 3, 4, 5} would only cover 3 paths out of the 5. So it seems like it works, but I don't yet have a universal correctness argument over all cases...

Reply:
OTOH if you have a query graph that looks like this:

~~~~
     3 -> 2 -> 1
          ^
     4 ---|
          |
     5 ---|
          |
7 -> 6 ---|
          |
9 -> 8 ---|
~~~~

Then intuitively the correct solution is {1, 2, 3, 4, 5}, since 7 and 9 are all vouching for 2 and 1, so it's OK for 3, 4, 5 to "vouch for" themselves. However this doesn't fit very well into the "hop" columns; previously I've been suggesting to simple ignore (delete) the 6 -> 2 and 8 -> 2 edges but that would collapse this example into the example in the previous post, but they have different intuitive answers...

Reply:
OK, the algorithm below is "obviously correct" but involves a bit of brute-force, it would be better to make it more efficient:

1. From the query graph, figure out what every node's sources are. So for example the graph in [this comment](https://github.com/libp2p/rust-libp2p/pull/1473#discussion_r423349277) becomes the following dictionary, which can be viewed as a graph G:

~~~~
V1 -> {3,4,5,7,9}
V2 -> {3,4,5}
V3 -> {3}
V4 -> {4}
V5 -> {5}
V6 -> {7}
V7 -> {7}
V8 -> {9}
V9 -> {9}
~~~~

The V prefix is just to distinguish e.g. V7 on the LHS, from 7 on the RHS, which makes the next step easier:

2. Iterate through all the d-combinations of the nodes, in sorted order. So for example this will be {V1,V2,V3,V4,V5}, {V1,V2,V3,V4,V6}, {V1,V2,V3,V4,V7}, etc. Take the subgraph of G induced by the combination, e.g. for {V1-V5} this is:

~~~~
V1 -> {3,4,5,7,9}
V2 -> {3,4,5}
V3 -> {3}
V4 -> {4}
V5 -> {5}
~~~~

Is there a [maximum bipartite matching](https://www.google.com/search?q=maximum+bipartite+matching)? If so, return the LHS as the set of results. If not, continue onto the next combination in the iteration.

The algorithm is guaranteed to terminate since there is always one matching {Vn -> {n} for all n in sources}, but I don't know if it's possible to construct pathological cases that take a long time...

Reply:
Instead of doing step (2) we can convert the whole thing into a single maximum-flow problem. First convert G from step (1) into a single flow graph as follows. For each source, connect it to each target with a flow of (1000000 - target node distance). (The number 1000000 is chosen just so the closest `d-1` targets have less flow than the furthest `d` targets; in the real network we might have to choose a bigger number.)

Then create a typical flow graph, with edges from a flow-source to all the sources, and from all the targets to a flow-sink. So the example above will be converted to:

~~~~
      (3) ---(1m - 1)--> V1
      ^                   \
     /                     \
  (1m)       [etc]         (1m)
   /         [etc]           \
  /                           v
 flow                        flow
source                       sink
  \                           ^
   \                         /
  (1m)                     (1m)
     \                     /
      v                   /
     (9) ----(1m - 9)--> V9
~~~~

Then the problem turns into a simple maximum-flow problem - the maximum flow will select d flows from the source to the sink, going through d of the Vn, with the minimal sum-over-n. Depending on [the algorithm used](https://en.wikipedia.org/wiki/Maximum_flow_problem#Solutions), the complexity is O(V^3) which should be fine for what we need.

Reply:
The above conversion is not exactly right, but I've figured out the correct version [here](https://forum.web3.foundation/t/s-kademlia-and-multipaths/310/7?u=infinity0) that also works to select the next-peer-to-query during the running of the algorithm.